File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/OrParser.scala
package jp.co.teruuu.parser.common

class OrParser[T](lps: Parser[T], rps: Parser[T]) extends Parser[T] {

  override def parse(input: String): ParseResult[T] = {
    lps.parse(input) match {
      case lResult: ParseOk[T] => ParseOk(lResult.value, lResult.next)
      case lResult: ParseNg[_] => {
        rps.parse(input) match {
          case rResult: ParseOk[T] => ParseOk(rResult.value, rResult.next)
          case rResult: ParseNg[_] => ParseNg(rResult.message, rResult.next)
        }
      }
    }
  }
}

File: java/parsers_java21/src/main/java/jp/co/teruuu/parser/common/OrParser.java
package jp.co.teruuu.parser.common;

import java.util.List;

public class OrParser<X> implements Parser<X> {
    private final Parser<X> fparser;
    private final Parser<X> sparser;

    public OrParser(Parser<X> fparser, Parser<X> sparser) {
        this.fparser = fparser;
        this.sparser = sparser;
    }

    @Override
    public ParseResult<X> parse(String input, int location) {
        switch (fparser.parse(input, location)) {
            case ParseResult.Success<X> fsuccess -> {
                return fsuccess;
            }
            case ParseResult.Failure<X> ffailure -> {
                switch (sparser.parse(input, location)) {
                    case ParseResult.Success<X> ssuccess -> {
                        return ssuccess;
                    }
                    case ParseResult.Failure<X> sfailure -> {
                        return new ParseResult.Failure<>(String.join(",", List.of(ffailure.message(), sfailure.message())), location);
                    }
                }
            }
        }
    }
}
